### ISSUE TITLE
model fails to load

### ISSUE THREAD

**Issue by bobcvn** [Dobiasd/frugally-deep#357]
Hello!
I have a multi-input model that I converted using  `convert_model.py` tool.

But it fails to load when I try to use it in my code.
The following code
```
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_model.json");
}
```
return the following error when executed:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  test failed: output=0 pos=0,0,0 value=-nan target=1
```
Here is a representation of the original model.
![model](https://user-images.githubusercontent.com/65460893/178030615-f63bb7fe-9052-43cd-898a-2b448246dac6.png)

The h5/json files are rather large.

Could you give me a hint about the issue?

Thank you for your time.

**Comment by Dobiasd** [comment]
Hi, and thanks for the very good report. :-)

The model architecture and used layer types look like something that normally should work in frugally-deep without any issue. Can you please try the following?
- Make sure you're using the latest versions of TensorFlow, Keras, and frugally-deep.
- Try the things described here: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#what-to-do-when-loading-my-model-with-frugally-deep-throws-an-stdruntime_error-with-test-failed

If these don't help, it would be cool if you could upload your (even if large) ˋ.h5ˋ file somewhere and send me the link. This way, I would be able to reproduce the error locally and debug into it. (starting from 2022-07-19, since I'm currently afk-ish)

**Comment by bobcvn** [comment]
Hi again and thank you for your suggestions!, I have tried them:
Keras version is 2.9.0. Tensorflow version is 2.9.1, frugally-deep is master branch version.

- Softening `verify_epsilon` or changing `FDEEP_FLOAT_TYPE` to double did not lead to better results. The model still won't load.
- When disabling the tests, the model will load but results in weird predictions on a dummy example: ` [[[[[[00nan, 00nan]]]]]]`. The result should be `[[0, 1]]`.

I uploaded the model file here:
https://wetransfer.com/downloads/66545c7cb940a53cd03114f4297e7fd320220719103240/ff1fbdec08e6092555e2a18bd65f921420220719103304/0c7457

Thank you again for your time.

**Comment by Dobiasd** [comment]
With your model file, I'm able to reproduce the problem locally ([`Dockerfile` for doing so](https://gist.github.com/Dobiasd/4246791c112ccdc9a00f33fcb733b589) -> [output](https://gist.github.com/Dobiasd/7cc04692d4a0093010bd2de3c9644035)).

I'll investigate and get back to you with the results.

**Comment by Dobiasd** [comment]
Using `feenableexcept(FE_ALL_EXCEPT & ~FE_INEXACT);`, I found it's `normalization_layer::apply_impl` doing a division by `0`.

![FIiMZsD](https://i.imgur.com/FIiMZsD.png)

The `0` comes from read access to `variance_[idx]` with `idx` being `2`, despite `variance_` only having 1 element. So it's not something with floating point precision, but a *real* bug. I'll keep you updated.

**Comment by Dobiasd** [comment]
Fixing that in Frugally-deep a generic way might be somewhat involved. As a quick workaround, you could try changing your input shapes from
- `(147, 251, 1)` (`Input_spec`)
- `(81, 1)` (`Input_srs`)

to
- `(147, 251)` (`Input_spec`)
- `(81)` (`Input_srs`)

if possible.

**Comment by bobcvn** [comment]
Thank you very much for your support.

**Comment by bobcvn** [comment]
Your suggestion works perfectly! 

For your information, the inputs dimension were chosen this way because we use convolutional layers and keras was returning an error when the flat inputs were used. Using "reshape layer" prevents this error (https://keras.io/api/layers/reshaping_layers/reshape/).

So now the model looks like this
![model](https://user-images.githubusercontent.com/65460893/180013525-6be4adff-c22f-4726-979e-b26fab72a181.png)

Thank you again!

**Comment by Dobiasd** [comment]
Great! :rocket: I'm happy to hear that. :relieved:

I'll leave this issue here open nonetheless until I have fixed the behavior of Frugally-deep's `Normalization` layer on these special input shapes. :construction_worker:

**Comment by Dobiasd** [comment]
@bobcvn Thanks again for helping me find (and thus fix) this bug in frugally-deep's implementation of the `Normalization` layer. :heart:

With the latest (just released) version (`v0.15.19-p0`), your original model (i.e., without the `Reshape` layer) works. :heavy_check_mark:

**Comment by bobcvn** [comment]
Thank **you** for your hard work. Glad it was helpful.
